### PR TITLE
[FIX] Recursive onchange bug

### PR DESCRIPTION
**Pull request by antespi**
@Yajo, @pedrobaeza this is another fix for the bug that @Yajo discovered from https://github.com/OCA/partner-contact/pull/104

Here I use `is_company == True` condition to cut recursion on `_inverse_name_after_cleaning_whitespace` method (called by `_compute_name > _onchange_name`) and use `is_company` on_change method to force call `_inverse_name` when `is_company` changes to `False`

Please, compare this fix with https://github.com/OCA/partner-contact/pull/139 and express your opinion

### CODE REVIEW

**Comment by gurneyalex** [comment]
@antespi: obvious difference is that #139 has a green build so my preference goes there at the moment.

File: partner_firstname/models.py [hunk 1]
```diff
@@ -62,7 +65,8 @@ def _inverse_name_after_cleaning_whitespace(self):
             self.name = clean
 
         # Save name in the real fields
-        else:
+        # elif self.is_company or self.env.context.get('force_inverse', False):
+        elif self.is_company:
```
Comment (yajo) [inline partner_firstname/models.py:69]: And what happens if the created partner is not a company?

File: partner_firstname/models.py [hunk 2]
```diff
@@ -62,7 +65,8 @@ def _inverse_name_after_cleaning_whitespace(self):
             self.name = clean
 
         # Save name in the real fields
-        else:
+        # elif self.is_company or self.env.context.get('force_inverse', False):
```
Comment (yajo) [inline partner_firstname/models.py:68]: This comment lies.

File: partner_firstname/models.py [hunk 3]
```diff
@@ -39,11 +40,13 @@ class ResPartner(models.Model):
...
 
     @api.one
-    @api.depends("firstname", "lastname")
+    @api.depends("firstname", "lastname", "company_id")
     def _compute_name(self):
         """Write the 'name' field according to splitted data."""
-        self.name = u" ".join((p for p in (self.lastname,
-                                           self.firstname) if p))
+        names = (self.lastname, self.firstname)
+        if self.company_id.names_order == 'first_last':
```
Comment (yajo) [inline partner_firstname/models.py:47]: Seems like you are merging the onchange fix with the new feature of choose names order. IMHO splitting that in 2 PR will make it easier to review.

**Comment by antespi** [comment]
@Yajo, @gurneyalex I'm preparing a rebase and squashed commit to this PR today for:
- Refactoring test suite, adding more exhaustive tests
- Removing old comments and debug line
- Rebase to upstream/8.0 for not including `company_id.names_order` feature, because it's a not merged one

Thanks for your reviews

**Comment by yajo** [comment]
Nice! I would suggest you to not lose too much time in fixing the onchange bug, since as says @gurneyalex in https://github.com/OCA/partner-contact/pull/143#issuecomment-122785071, it's already fixed and well tested in https://github.com/OCA/partner-contact/pull/139 (and it was indeed very hard to fix).

You could also consider putting the names order feature in a separate module, but it's a matter of choice.

**Comment by antespi** [comment]
@gurneyalex, @Yajo I've

**Comment by antespi** [comment]
@gurneyalex, @Yajo I've close this PR because i've preferred to continue from @Yajo PR (https://github.com/OCA/partner-contact/pull/139) and I've just propuse to him this one (https://github.com/grupoesoc/partner-contact/pull/2), with a new test suite for covering more cases and fixes to pass them.
Please, review them and merge it if everything is OK. All tests passed in my local environment

**Pull request closed without merging**